File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func hasCycle(head *ListNode) bool {
    // create map of values
    // if the next value gives the value already in the list
    // return true

    visited := make(map[*ListNode]bool)

    for head != nil {
        if visited[head] {
            return true
        } else {
            visited[head] = true
            head = head.Next
        }
    }
    return false
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        
        if not head:
            return None
        
        dummy = prev = ListNode(0, head)
        curr = head

        while curr:
            count = 1

            while curr and count < k:
                curr = curr.next
                count += 1
            
            if not curr:
                break
            
            nprev, ncurr = prev.next, curr.next
            check = curr

            while prev.next != check:
                node = prev.next
                prev.next = node.next
                node.next = curr.next
                curr.next = node

            prev, curr = nprev, ncurr

        return dummy.next
            
            